File: dotnet/Lexy/Lexy.Compiler/Language/Expressions/FunctionCallExpression.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.Language.Expressions.Functions;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;

namespace Lexy.Compiler.Language.Expressions
{
    public class FunctionCallExpression : Expression
    {
        public string FunctionName { get; }

        public List<Expression> Arguments { get; }
        public BuiltInFunction BuiltInFunction { get; }

        private FunctionCallExpression(string functionName, List<Expression> arguments,
            BuiltInFunction builtInFunction,
            ExpressionSource source, SourceReference reference) : base(source, reference)
        {
            FunctionName = functionName ?? throw new ArgumentNullException(nameof(functionName));
            Arguments = arguments;
            BuiltInFunction = builtInFunction;
        }

        public static ParseExpressionResult Parse(ExpressionSource source)
        {
            var tokens = source.Tokens;
            if (!IsValid(tokens))
            {
                return ParseExpressionResult.Invalid<FunctionCallExpression>("Not valid.");
            }

            var matchingClosingParenthesis = ParenthesizedExpression.FindMatchingClosingParenthesis(tokens);
            if (matchingClosingParenthesis == -1)
            {
                return ParseExpressionResult.Invalid<FunctionCallExpression>("No closing parentheses found.");
            }

            var functionName = tokens.TokenValue(0);
            var innerExpressionTokens = tokens.TokensRange(2, matchingClosingParenthesis - 1);
            var argumentsTokenList = ArgumentList.Parse(innerExpressionTokens);
            if (!argumentsTokenList.IsSuccess)
            {
                return ParseExpressionResult.Invalid<FunctionCallExpression>(argumentsTokenList.ErrorMessage);
            }

            var arguments = new List<Expression>();
            foreach (var argumentTokens in argumentsTokenList.Result)
            {
                var argumentExpression = ExpressionFactory.Parse(source.File, argumentTokens, source.Line);
                if (argumentExpression.Status == ParseExpressionStatus.Failed) return argumentExpression;

                arguments.Add(argumentExpression.Expression);
            }

            var reference = source.CreateReference();

            var builtInFunctionResult = BuiltInFunctions.Parse(functionName, source.CreateReference(), arguments);
            if (!builtInFunctionResult.IsSuccess)
            {
                return ParseExpressionResult.Invalid<FunctionCallExpression>(builtInFunctionResult.ErrorMessage);
            }

            var expression = new FunctionCallExpression(functionName, arguments, builtInFunctionResult.Result, source, reference);

            return ParseExpressionResult.Success(expression);
        }

        public static bool IsValid(TokenList tokens)
        {
            return tokens.IsTokenType<StringLiteralToken>(0)
                   && tokens.OperatorToken(1, OperatorType.OpenParentheses);
        }

        public override IEnumerable<INode> GetChildren()
        {
            if (BuiltInFunction != null)
            {
                yield return BuiltInFunction;
            }
        }

        protected override void Validate(IValidationContext context)
        {
            if (BuiltInFunction == null)
            {
                context.Logger.Fail(Reference, $"Invalid function name: '{FunctionName}'");
                return;
            }
        }

        public override VariableType DeriveType(IValidationContext context)
        {
            return BuiltInFunction?.DeriveReturnType(context);
        }
    }
}

File: typescipt/src/language/expressions/functionCallExpression.ts
import {Expression} from "./Expression";

export class FunctionCallExpression extends Expression {


  public nodeType: "FunctionCallExpression"

  public string FunctionName

   public Array<Expression> Arguments
   public ExpressionFunction ExpressionFunction

   private FunctionCallExpression(string functionName, Array<Expression> arguments,
     ExpressionFunction expressionFunction,
     ExpressionSource source, SourceReference reference) {
    {super(source, reference) {
     FunctionName = functionName ?? throw new Error(nameof(functionName));
     Arguments = arguments;
     ExpressionFunction = expressionFunction;
   }

   public static parse(source: ExpressionSource): ParseExpressionResult {
     let tokens = source.tokens;
     if (!IsValid(tokens)) return newParseExpressionFailed(FunctionCallExpression>(`Not valid.`);

     let matchingClosingParenthesis = ParenthesizedExpression.FindMatchingClosingParenthesis(tokens);
     if (matchingClosingParenthesis == -1)
       return newParseExpressionFailed(FunctionCallExpression>(`No closing parentheses found.`);

     let functionName = tokens.tokenValue(0);
     let innerExpressionTokens = tokens.tokensRange(2, matchingClosingParenthesis - 1);
     let argumentsTokenList = ArgumentList.parse(innerExpressionTokens);
     if (!argumentsTokenList.state != 'success')
       return newParseExpressionFailed(FunctionCallExpression>(argumentsTokenList.errorMessage);

     let arguments = new Array<Expression>();
     foreach (let argumentTokens in argumentsTokenList.result) {
       let argumentExpression = ExpressionFactory.parse(argumentTokens, source.line);
       if (!argumentExpression.state != 'success') return argumentExpression;

       arguments.Add(argumentExpression.result);
     }

     let reference = source.createReference();

     let builtInFunctionResult = BuiltInExpressionFunctions.parse(functionName, source.createReference(), arguments);
     if (builtInFunctionResult is { IsSuccess: false })
       return newParseExpressionFailed(FunctionCallExpression>(builtInFunctionResult.errorMessage);

     let expressionFunction = builtInFunctionResult?.result
                 ?? new LexyFunction(functionName, arguments, source.createReference());

     let expression = new FunctionCallExpression(functionName, arguments, expressionFunction, source, reference);

     return newParseExpressionSuccess(expression);
   }

   public static isValid(tokens: TokenList): boolean {
     return tokens.isTokenType<StringLiteralToken>(0)
        && tokens.operatorToken(1, OperatorType.OpenParentheses);
   }

   public override getChildren(): Array<INode> {
     if (ExpressionFunction != null) yield return ExpressionFunction;
   }

   protected override validate(context: IValidationContext): void {
     if (ExpressionFunction == null) {
       context.logger.fail(this.reference, $`Invalid function name: '{FunctionName}'`);
     }
   }

   public override deriveType(context: IValidationContext): VariableType {
     return ExpressionFunction?.DeriveReturnType(context);
   }
}
